src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala

```diff
@@ -195,32 +196,28 @@ private[internal] trait TypeConstraints {
 
   /** Solve constraint collected in types `tvars`.
     *
-    *  @param tvars      All type variables to be instantiated.
-    *  @param tparams    The type parameters corresponding to `tvars`
-    *  @param variances  The variances of type parameters; need to reverse
+    *  @param tvars    All type variables to be instantiated.
+    *  @param tparams  The type parameters corresponding to `tvars`
+    *  @param getVariance Function to extract variances of type parameters; we need to reverse
     *                    solution direction for all contravariant variables.
-    *  @param upper      When `true` search for max solution else min.
+    *  @param upper    When `true` search for max solution else min.
     */
-  def solve(tvars: List[TypeVar], tparams: List[Symbol], variances: List[Variance], upper: Boolean, depth: Depth): Boolean = {
-    // We're going to be iterating over these a lot. Let's put them in an array.
```

Comment:
This merge goes back to iterating `List`s, however we now just iterate-with-index over the `tvars` list, use the index to lookup the variance in the `BitSet`, and follow `TypeVar.origin.typeSymbol` to get the corresponding `tparams`.

I'll keep an eye on this code path in my allocation/CPU profiles.

Reply:
After this merge:

![image](https://user-images.githubusercontent.com/65551/55061985-c0cb4280-50c0-11e9-8974-33f9d6a5b7aa.png)

The time in solve is largely spent extracting the variances, calling `lub` and `=:=`

Reply:
@retronym I am not sure I understand the graph. Is the forward-port improving or hurting performance? We can still add the array thing, if we can see it helps.

Reply:
That profile tells me there the iteration time of the list of tvars is neglgible -- there is very little "self time" in `solve`/`solveOne`, and very little of its nested time is spend in `List.{head,tail}`. These lists are typically very small.

If you want to view these profiles:
  - download collapsed stacks from a recent profile run: https://scala-ci.typesafe.com/view/scala-bench/job/compiler-benchmark/2539/artifact/target/profile-async/collapsed-cpu.txt
  - Open that file in https://www.speedscope.app/
  - I'm using ['Sandwich'](https://github.com/jlfwong/speedscope#-sandwich) mode there to focus on a method of interest.